instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module AwsSdk
+      # Generates Spans for all interactions with AwsSdk
+      class Handler < Seahorse::Client::Handler
+        def call(context)
+          span_name, attributes = calculate_span(context)
+
+          tracer.in_span(span_name, kind: OpenTelemetry::Trace::SpanKind::CLIENT, attributes: attributes) do |span|
+            execute = proc {
+              super(context).tap do |response|
+                if (err = response.error)
+                  span.record_exception(err)
+                  span.status = Trace::Status.error(err)
+                end
+              end
+            }
+
+            if instrumentation_config[:suppress_internal_instrumentation]
+              OpenTelemetry::Common::Utilities.untraced(&execute)
+            else
+              execute.call
+            end
+          end
+        end
+
+        def calculate_span(context)
+          service_name = context.client.class.api.metadata['serviceId'] || context.client.class.to_s.split('::')[1]
```

Comment:
I used [xray as inspiration](https://github.com/aws/aws-xray-sdk-ruby/blob/master/lib/aws-xray-sdk/facets/aws_sdk.rb#L21-L22) for service name extraction.
Added safe navigation in [this commit](https://github.com/open-telemetry/opentelemetry-ruby/pull/1014/commits/31ed27fa58aa9a42a45fc1289409caac506e87d6#diff-1723a57ddb33bc724a6da2be5c6a3dffa0a655cb65a43fbe00a78c71f5d2cb89R44).

---

I tried to use safe navigation for every method in the chain, but I got error from `rubocop`. 
@mwear What is the best practice in this case?

### What I tried:
```ruby
def get_service_name(context)
  context&.client&.class&.api&.metadata['serviceId'] || context&.client&.class&.to_s&.split('::')[1]
end
```

### The warning I get:
> lib/opentelemetry/instrumentation/aws_sdk/handler.rb:44:48: W: Lint/SafeNavigationChain: Do not chain ordinary method call after safe navigation operator.
>           context&.client&.class&.api&.metadata['serviceId'] || context&.client&.class&.to_s&.split('::')[1]

Reply:
chaining after `.to_s` is unnecessary in this case, and likely  `.class` as well. basically if the class exists then it would be reasonable to assume `.class` would return as that method exists on every class, and likewise with `.to_s`. Does that help?

Reply:
Thanks @ericmustin!

Can you confirm this chaining is safe?
```rb
context&.client.class.api.metadata['serviceId'] || context&.client.class.to_s.split('::')[1]
```

Reply:
I have no idea off hand, i'd have to look into the implementation 🙃. I can follow up when i have some more bandwidth. 

specifically, are `client`, `api` and `metadata` always guaranteed to return what you expect here? If it's possible they return `nil` then we'd want to chain those with the safe operator

Reply:
@ericmustin I was referring to this. In the case context is `nil` it will always results in a `NoMethodError` because `NilClass` does not have a class method `api`. 

```
omoplata ~(:|✔) %
🤘 irb
irb(main):001:0> context = nil
=> nil
irb(main):002:0> context&.client.api
Traceback (most recent call last):
        4: from /Users/arielvalentin/.rbenv/versions/2.7.4/bin/irb:23:in `<main>'
        3: from /Users/arielvalentin/.rbenv/versions/2.7.4/bin/irb:23:in `load'
        2: from /Users/arielvalentin/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        1: from (irb):2
NoMethodError (undefined method `api' for nil:NilClass)
```

Reply:
ah far as i can tell, `context`, `client`, and `api` should be initialized with default values and never return `nil`

- `context` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/client/response.rb#L13

- `client` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/client/request_context.rb#L39-L40

- `api` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/client/base.rb#L171-L173

- `metadata` https://github.com/aws/aws-sdk-ruby/blob/b5c46409986c6681b35533e017051d55e733d053/gems/aws-sdk-core/lib/seahorse/model/api.rb#L8

Technically I guess someone could provide a junk value for context, but many other parts of the aws-sdk would break in that situation anyway, so i'm not too concerned.

tl;dr, i think this is much ado about nothing, 

`context.client.class.api.metadata['serviceId'] || context.client.class.to_s.split('::')[1]` is fine imo